src/main/java/com/conveyal/gtfs/loader/ReferenceTracker.java

```diff
@@ -136,4 +150,120 @@
+                        //  data, but the first doesn't define an agency_id.
+                        String entityId = getValueForRow(rowData, table.getKeyFieldIndex(fields));
+                        String message = String.format(
+                            "%s is conditionally required when there is more than one agency.",
+                            check.conditionalFieldName
+                        );
+                        errors.add(
+                            NewGTFSError.forLine(table, lineNumber, AGENCY_ID_REQUIRED_FOR_TABLES_WITH_MORE_THAN_ONE_RECORD, message).setEntityId(entityId)
+                        );
+                    }
+                } else if (check.referenceCheck == FIELD_IN_RANGE) {
+                    boolean referenceValueMeetsRangeCondition =
+                        !POSTGRES_NULL_TEXT.equals(refFieldData) &&
+                        // TODO use pre-existing method in ShortField?
+                        isValueInRange(refFieldData, check.minReferenceValue, check.maxReferenceValue);
+                    // If ref value does not meet the range condition, there is no need to check this conditional value for
+                    // (e.g.) an empty value. Continue to the next check.
+                    if (!referenceValueMeetsRangeCondition) continue;
+                    int conditionalFieldIndex = Field.getFieldIndex(fields, check.conditionalFieldName);
+                    String conditionalFieldData = getValueForRow(rowData, conditionalFieldIndex);
+                    boolean conditionallyRequiredValueIsEmpty = check.conditionalCheck == FIELD_NOT_EMPTY &&
+                        POSTGRES_NULL_TEXT.equals(conditionalFieldData);
+                    if (conditionallyRequiredValueIsEmpty) {
+                        String entityId = getValueForRow(rowData, table.getKeyFieldIndex(fields));
+                        String message = String.format(
+                            "%s is conditionally required when %s value is between %d and %d.",
+                            check.conditionalFieldName,
+                            referenceField.name,
+                            check.minReferenceValue,
+                            check.maxReferenceValue
+                        );
+                        errors.add(
+                            NewGTFSError.forLine(table, lineNumber, CONDITIONALLY_REQUIRED, message).setEntityId(entityId)
+                        );
+                    }
+                } else if (check.referenceCheck == FOREIGN_FIELD_VALUE_MATCH) {
+                    String foreignFieldReference = String.join(":", check.conditionalFieldName, refFieldData);
+                    if (table.name.equals("fare_rules") &&
+                        !POSTGRES_NULL_TEXT.equals(refFieldData) &&
+                        foreignFieldIds.stream().noneMatch(id -> id.contains(foreignFieldReference))
```

Comment:
Since `foreignFieldIds` is a set and the `foreignFieldReference` seems to be the actual value that we're searching for, it seems like it'd be possible to do the following:

```suggestion
                        !foreignFieldIds.contains(foreignFieldReference)
```

otherwise, this is needlessly iterating through potentially the entire set of `foreignFieldIds`.

Reply:
Will be using this for agency_id so leaving as is.